test/functional/test_framework/ellsq.py

```diff
@@ -0,0 +1,243 @@
+# Source: https://github.com/sipa/writeups/tree/main/elligator-square-for-bn
+"""Test-only Elligator Squared implementation
+
+WARNING: This code is slow and uses bad randomness.
+Do not use for anything but tests."""
+
+import random
+import secrets
+import unittest
+
+from .key import fe, SECP256K1, SECP256K1_FIELD_SIZE, SECP256K1_G, SECP256K1_ORDER
+
+C1 = fe(-3).sqrt()
+C2 = (C1 - fe(1)) / fe(2)
+B = fe(7)
+
+def f(u):
+    """Forward mapping function"""
+    s = u**2
+    x1 = C2 - C1*s / (fe(1)+B+s)
+    g1 = x1**3 + B
+    if g1.is_square():
+        x, g = x1, g1
+    else:
+        x2 = -x1 - fe(1)
+        g2 = x2**3 + B
+        if g2.is_square():
+            x, g = x2, g2
+        else:
+            x3 = fe(1) - (fe(1)+B+s)**2 / (fe(3)*s)
+            g3 = x3**3 + B
+            x, g = x3, g3
+    y = g.sqrt()
+    if y.is_odd() == u.is_odd():
+        return (x, y)
+    else:
+        return (x, -y)
+
+def r(x,y,i):
+    """Reverse mapping function"""
```

Comment:
Same here, something like "Given the X and Y coordinates of a point on the secp256k1 curve (of type fe), return a value u such that f(u) = (x,y), or None. There can be up to 4 such inverses, and i selects which formula to use. Each i can independently from other i values return a value or None. All non-None values returned across all 4 i values are guaranteed to be distinct, and together they will cover all inverses of (x,y) under f."

Reply:
Thanks! Done.